pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -895,16 +904,27 @@ public void redeliverUnacknowledgedMessages() {
     }
 
     @Override
-    public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
+    public void redeliverUnacknowledgedMessages(Set<MessageIdImpl> messageIds) {
```

Comment:
yes, that was a thought to avoid ```remove``` in case of hole. In that case: poll(0,ms) N msgs from queue can fetch non-timeout messaged if client consumes timeout msg from the queue.